File: node/ts/zkgroup/calllinks/CreateCallLinkCredential.ts
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { randomBytes } from 'node:crypto';

import ByteArray from '../internal/ByteArray.js';
import { RANDOM_LENGTH } from '../internal/Constants.js';
import * as Native from '../../Native.js';

import CallLinkSecretParams from './CallLinkSecretParams.js';
import CreateCallLinkCredentialPresentation from './CreateCallLinkCredentialPresentation.js';
import GenericServerPublicParams from '../GenericServerPublicParams.js';
import { Aci } from '../../Address.js';

export default class CreateCallLinkCredential extends ByteArray {
  private readonly __type?: never;

  constructor(contents: Uint8Array) {
    super(contents, Native.CreateCallLinkCredential_CheckValidContents);
  }

  present(
    roomId: Uint8Array,
    userId: Aci,
    serverParams: GenericServerPublicParams,
    callLinkParams: CallLinkSecretParams
  ): CreateCallLinkCredentialPresentation {
    const random = randomBytes(RANDOM_LENGTH);
    return this.presentWithRandom(
      roomId,
      userId,
      serverParams,
      callLinkParams,
      random
    );
  }

  presentWithRandom(
    roomId: Uint8Array,
    userId: Aci,
    serverParams: GenericServerPublicParams,
    callLinkParams: CallLinkSecretParams,
    random: Uint8Array
  ): CreateCallLinkCredentialPresentation {
    return new CreateCallLinkCredentialPresentation(
      Native.CreateCallLinkCredential_PresentDeterministic(
        this.contents,
        roomId,
        userId.getServiceIdFixedWidthBinary(),
        serverParams.contents,
        callLinkParams.contents,
        random
      )
    );
  }
}

File: java/shared/java/org/signal/libsignal/zkgroup/calllinks/CreateCallLinkCredential.java
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.zkgroup.calllinks;

import static org.signal.libsignal.internal.FilterExceptions.filterExceptions;
import static org.signal.libsignal.zkgroup.internal.Constants.RANDOM_LENGTH;

import java.security.SecureRandom;
import org.signal.libsignal.internal.Native;
import org.signal.libsignal.protocol.ServiceId.Aci;
import org.signal.libsignal.zkgroup.GenericServerPublicParams;
import org.signal.libsignal.zkgroup.InvalidInputException;
import org.signal.libsignal.zkgroup.internal.ByteArray;

public final class CreateCallLinkCredential extends ByteArray {

  public CreateCallLinkCredential(byte[] contents) throws InvalidInputException {
    super(contents);
    filterExceptions(
        InvalidInputException.class,
        () -> Native.CreateCallLinkCredential_CheckValidContents(contents));
  }

  public CreateCallLinkCredentialPresentation present(
      byte[] roomId,
      Aci userId,
      GenericServerPublicParams serverParams,
      CallLinkSecretParams callLinkParams) {
    return present(roomId, userId, serverParams, callLinkParams, new SecureRandom());
  }

  public CreateCallLinkCredentialPresentation present(
      byte[] roomId,
      Aci userId,
      GenericServerPublicParams serverParams,
      CallLinkSecretParams callLinkParams,
      SecureRandom secureRandom) {
    byte[] random = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents =
        filterExceptions(
            () ->
                Native.CreateCallLinkCredential_PresentDeterministic(
                    getInternalContentsForJNI(),
                    roomId,
                    userId.toServiceIdFixedWidthBinary(),
                    serverParams.getInternalContentsForJNI(),
                    callLinkParams.getInternalContentsForJNI(),
                    random));

    try {
      return new CreateCallLinkCredentialPresentation(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }
}
